packages/multichain/src/scope/validation.ts

```diff
@@ -0,0 +1,129 @@
+import { isCaipReference } from '@metamask/utils';
+
+import type {
+  ExternalScopeString,
+  ExternalScopeObject,
+  ExternalScopesObject,
+} from './types';
+import { parseScopeString } from './types';
+
+/**
+ * Validates a scope object according to the [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) spec.
+ * @param scopeString - The scope string to validate.
+ * @param scopeObject - The scope object to validate.
+ * @returns A boolean indicating if the scope object is valid according to the [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) spec.
+ */
+export const isValidScope = (
+  scopeString: ExternalScopeString,
+  scopeObject: ExternalScopeObject,
+): boolean => {
+  const { namespace, reference } = parseScopeString(scopeString);
+
+  // Namespace is required
+  if (!namespace) {
+    return false;
+  }
+
+  const {
+    references,
+    methods,
+    notifications,
+    accounts,
+    rpcDocuments,
+    rpcEndpoints,
+    ...extraProperties
+  } = scopeObject;
+
+  // Methods and notifications are required
+  if (!methods || !notifications) {
+    return false;
+  }
+
+  // For namespaces other than 'wallet', either reference or non-empty references array must be present
+  if (
+    namespace !== 'wallet' &&
+    !reference &&
+    (!references || references.length === 0)
+  ) {
+    return false;
+  }
+
+  // If references are present, reference must be absent and all references must be valid
+  if (references) {
+    if (reference && references.length > 0) {
+      return false;
+    }
+
+    const areReferencesValid = references.every((nestedReference) =>
+      isCaipReference(nestedReference),
+    );
+
+    if (!areReferencesValid) {
+      return false;
+    }
+  }
+
+  const areMethodsValid = methods.every(
+    (method) => typeof method === 'string' && method.trim() !== '',
+  );
+
+  if (!areMethodsValid) {
+    return false;
+  }
+
+  const areNotificationsValid = notifications.every(
+    (notification) =>
+      typeof notification === 'string' && notification.trim() !== '',
+  );
+
+  if (!areNotificationsValid) {
+    return false;
+  }
+
+  // Ensure no unexpected properties are present in the scope object
+  if (Object.keys(extraProperties).length > 0) {
+    return false;
+  }
+
+  return true;
+};
+
+/**
+ * Validates a set of scopes according to the [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) spec.
+ * @param requiredScopes - The required scopes to validate.
+ * @param optionalScopes - The optional scopes to validate.
+ * @returns An object containing the valid required scopes and optional scopes.
+ */
+export const validateScopes = (
```

Comment:
This seems to filter out invalid scopes, rather than validate what's given. My expectation when reading this name and description was that it would throw an error if an invalid scope was encountered, rather than omitting it.

Maybe we could rename this to `getValidScopes`, and update the description to make clear that we're getting whatever subset is valid rather than validating what's passed in as a whole?

Reply:
done here:[ ](https://github.com/MetaMask/core/pull/4784/commits/b700e4f2398d0067519e12a3d8f2bc3f7e4e98c0)